apps/prairielearn/elements/pl-file-preview/pl-file-preview.js

```diff
@@ -110,15 +116,51 @@
               const type = blob.type;
               if (type === 'text/plain') {
                 const text = await blob.text();
-                code.textContent = text;
-                pre.classList.remove('d-none');
+                if (escapedFileName.endsWith('.ipynb')) {
+                  // importing the notebookjs library doesn't return an object, it sets the global variable 'ns'
+                  // importing DOMPurify sets the global variable DOMPurify.
+                  await Promise.all([
+                    import('marked'),
+                    import('dompurify'),
+                    import('notebookjs'),
+                  ]).then(async ([Marked]) => {
+                    /* Showdown has a small bug that doesn't allow it to be loaded dynamically.
+                       This PR will fix it: https://github.com/showdownjs/showdown/pull/1017
+                       Since the PR could take two weeks or two months, let's used Marked for now
+                       and get this feature deployed.*/
+                    nb.markdown = Marked.marked.parse;
+                    // const converter = new showdown.Converter();
+                    // nb.markdown = (text) => converter.makeHtml(text);
+
+                    nb.sanitizer = (code) =>
+                      DOMPurify.sanitize(code, { SANITIZE_NAMED_PROPS: true });
+                    const notebook = nb.parse(JSON.parse(text));
+                    const rendered = notebook.render();
+
+                    notebookPreview.appendChild(rendered);
+                    notebookPreview.classList.remove('d-none');
+
+                    // Typeset any math that might be in the notebook.
+                    window.MathJax.typesetPromise();
+                  });
+                } else {
+                  code.textContent = text;
+                  pre.classList.remove('d-none');
+                }
+
                 hideErrorMessage();
 
                 // Only show the expand/collapse button if the content is tall
-                // enough where scrolling is necessary.
-                if (pre.scrollHeight > pre.clientHeight) {
+                // enough where scrolling is necessary. This must be done before
+                // auto-expansion happens below.
+                if (container.scrollHeight > container.clientHeight) {
                   expandButton.classList.remove('d-none');
                 }
+
+                // Always fully expand notebook previews.
+                if (escapedFileName.endsWith('.ipynb')) {
+                  toggleExpanded(true);
+                }
```

Comment:
This is opinionated, but I think it's the right thing to do. A 200px-tall preview of a notebook is almost never going to be what someone wants.